doc/newpath.1
.\" $Id: newpath.1,v 1.8 2025/09/10 20:26:42 tom Exp $
.TH NEWPATH 1 2025-09-10 MiscTools "User commands"
.hy 0
.SH NAME
newpath \-
review or update the PATH
.SH SYNOPSIS
.B newpath
[ \fIoptions\fP [ \fIdirectories\fP ]]
[ - \fIcommand\fP ]
.
.SH DESCRIPTION
Echoes the \fBPATH\fP environment variable with the given directories added,
(the default) or removed.
This program was written in June 1994, and is used in a variety
of shell environments.
.PP
After the options and directory-parameters,
you can put a '\-' before a command
to invoke it with the environment variable updated,
rather than echoing the result to standard output.
.PP
When adding directories, \fBnewpath\fP resolves symbolic links
using the \fBrealpath\fP(3) function.
.
.SH OPTIONS
.TP 5
.B \-0
prepend the pathname where \fBnewpath\fP is found.
.TP 5
.BI \-a NAME
modify path after given \fINAME\fP.
.TP 5
.B \-b
put new arguments before existing path
.TP 5
.B \-d
remove duplicates/non-directory items
.TP 5
.B \-e
put new arguments after end of path
.TP 5
.B \-f
allow filenames to match, as well as directories
.TP 5
.BI \-n NAME
specify environment-variable to use (default: PATH)
.TP 5
.B \-p
print in C-shell form
.TP 5
.B \-r
remove arguments from path
.TP 5
.B \-v
verbose, shows the path with one directory per line.
.
.
.SH BUGS
.
Probably.
.
.SH EXAMPLES
.
\fBNewpath\fR is used in the shell initialization script to
set up customized PATH (or PATH-like) environment variables
which correspond to the directories which exist.
For example, it is used in systems that have NFS-mounted
home directories with different application directories on each host.
.PP
Because some of the utilities in this collection have the same name
as a few other programs,
it is helpful to install the collection in a directory
normally not on the PATH,
and (with a symbolic link to \fBnewpath\fP in a directory on the PATH),
use \fBnewpath\fP to augment the path to include those utilities.
.PP
This C shell script checks for the link, and uses it:
.sp
.RS
.EX
foreach dir ( /usr/bin /usr/local/bin )
    if ( -f $dir/newpath ) then
        set path=(`newpath -0dp`)
        break
    endif
end
.EE
.RE
.PP
If the link is assumed, either C or POSIX shell can augment the path
in a single line:
.sp
.RS
.EX
set path=(`newpath -0dp`)
PATH=`newpath -0d`
.EE
.RE
.
.SH SEE ALSO
sh(1)


CHANGES
2025-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/sue/sue.c: workaround for broken NetBSD header

	* td_getopt.h: NetBSD lint

	* package/debian/changelog, package/misc_tools.spec: bump

2025-09-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* COPYING.asc: sign

	* package/misc_tools.spec: SPDX equivalent

	* COPYING: consolidate license

	* install-bin.in:
	amend $INSTALL in case install-sh is used from a subdirectory

	* install-bin.in: shellcheck warnings

	* config_h.in, doc/minstall.in, package/debian/postinst,
	  src/newpath/makefile.wnt, configure.in:
	trim-blanks

	* src/unmap/chrcount.c, src/unmap/map.c, src/unmap/map_s.c,
	  src/unmap/unmap.c, src/unmap/unmap_s.c, src/width/width.c,
	  src/splitit/splitit.c, src/perror/perror.c, src/slowcat/slowcat.c,
	  src/sue/sue.c, src/tisplit/timerge.c, src/tisplit/tisplit.c,
	  src/acsplit/acmerge.c, src/count_files/count_files.c, src/cpd/cpd.c,
	  src/newest/newest.c, src/newpath/newpath.c, src/acsplit/acsplit.c,
	  td_getline.h:
	s/0/NULL/

	* doc/newpath.1: bump

	* doc/newpath.1: grammar

	* doc/cpd.1: bump

	* doc/cpd.1: grammar

	* doc/chrcount.1: bump

	* doc/chrcount.1: grammar

	* doc/acsplit.1: bump

	* doc/acsplit.1: grammar

	* package/debian/copyright: bump

	* configure: regen

	* package/debian/changelog, package/misc_tools.spec: bump

2025-08-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: update to 2024-07-27

2024-07-15  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-bin.in: source parameter may have $DESTDIR prefix

	* src/sue/use-usual:
	use $DestDir rather than $DESTDIR, and use the install program
	to install the prepared wrapper rather than fixing after install.

	* package/misc_tools.spec: use --with-execdir option

	* src/sue/use-hacks: quoting per shellcheck warnings
	do not exit on the first error, to work with rpmbuild

	* package/debian/changelog: bump

2024-07-13  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/newpath/newpath.c:
	a relative link for newpath, i.e., making /usr/bin/newpath a symbolic link
	to ../libexec/misctools/newpath did not work -- fix this by finding the
	program on the path, and using realpath to resolve links in that.

	* install-bin.in: corrected link source

	* package/debian/postinst: using libexec now, rather than lib

	* src/sue/use-hacks:
	s/DESTDIR/DestDir/ to avoid conflict with $DESTDIR used in package builds

	* configure: regen

	* src/sue/use-hacks: exit on the first error

	* install-bin.in: add special case for src/sue

	* configure.in: add install_bin symbol, for special case in src/sue

	* src/sue/makefile.in:
	bin-directory depends upon the --with-execdir option

	* configure: regen

	* install-bin.in: account for DESTDIR

	* configure.in: improve message-result for logging

	* package/debian/postinst:
	symbolic link is now done via configure option --with-execdir

	* package/debian/rules: use --with-execdir option

	* configure: regen

	* aclocal.m4: add CF_PROG_LN_S

	* install-bin.in: RCS_BASE

	* package/misc_tools.spec:
	simplify by adding a configure option --with-execdir

	* configure.in, src/acsplit/makefile.in, src/count_files/makefile.in,
	  src/cpd/makefile.in, src/hex/makefile.in, src/isatty/makefile.in,
	  src/newest/makefile.in, src/newpath/makefile.in,
	  src/perror/makefile.in, src/realpath/makefile.in,
	  src/slowcat/makefile.in, src/splitit/makefile.in, src/sue/makefile.in,
	  src/tisplit/makefile.in, src/unmap/makefile.in, src/width/makefile.in:
	added --with-execdir option to simplify installing the tools into libexecdir
	while making a symbolic link to newpath

	* makefile.in: provide for cleanup of install-bin script

	* package/misc_tools.spec:
	build all programs, since there is no naming conflict now

	* doc/minstall.in: update to match unquoted section number in .TH line

	* package/debian/changelog, package/misc_tools.spec: bump

2024-07-12  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/misc_tools.spec:
	add relative symlink for newpath to the %files list

	* doc/newpath.1:
	explain how to reduce environment conflicts with this set of utilities

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	quoting

	* doc/newpath.1: identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/newpath.1, doc/perror.1, doc/realpath.1, doc/slowcat.1,
	  doc/splitit.1, doc/sue.1, doc/timerge.1, doc/tisplit.1, doc/unmap.1,
	  doc/width.1:
	check-manpage

	* package/debian/copyright: bump

	* configure: regen

	* COPYING.asc: sign

	* COPYING, package/debian/changelog, package/misc_tools.spec: bump

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: update to 2024-05-27

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023/12/27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023/12/2

2023-12-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-04-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* doc/minstall.in: reduce filename-conflicts by installing ".1mt" files

	* configure: regen


src/newpath/newpath.c
/*
 * $Id: newpath.c,v 1.20 2025/09/10 20:38:30 tom Exp $
 *
 * Author:	T.E.Dickey
 * Created:	02 Jun 1994
 * Modified:
 *		13 Jul 2024, handle relative-link in search for original path.
 *		25 Mar 2021, add "-0" option.
 *		13 Mar 2012, integrate into misc_tools package
 *		10 Nov 2000, port to win32
 *
 * Function: (see usage message)
 */
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>

#ifdef WIN32
#include <process.h>
#else
#include <limits.h>
#include <unistd.h>
#endif

#include <td_getopt.h>

#ifdef WIN32
#define	PATHDELIM ';'
#define	PATHSEP   '\\'
#define Compare(a,b) stricmp(a,b)
#define SameDir(a,b) (!Compare((a)->nn, (b)->nn))
#else
#define	PATHDELIM ':'
#define	PATHSEP   '/'
#define Compare(a,b) strcmp(a,b)
#define SameDir(a,b) ((a)->sb.st_ino == (b)->sb.st_ino \
				   && (a)->sb.st_dev == (b)->sb.st_dev)
#endif

#ifndef	TRUE
#define	TRUE	1
#define	FALSE	0
#endif

#ifndef	EXIT_SUCCESS
#define	EXIT_SUCCESS 0
#define EXIT_FAILURE 1
#endif

#define	EOS	'\0'

#define	BLANK	""

#ifdef OPT_TRACE
#define	TRACE(p)				fprintf p
#else
#define	TRACE(p)		/* fprintf p; */
#endif

typedef struct {
    char *nn;
    struct stat sb;
} LIST;

static int allow_files;

static void
failed(const char *s)
{
    perror(s);
    exit(EXIT_FAILURE);
}

static void
usage(void)
{
    static const char *tbl[] =
    {
	"Usage: newpath [options [directories]] [ - command]",
	"",
	"Echos the PATH environment variable with the given directories added",
	"(the default) or removed.",
	"",
	"Options:",
	"    -0      put newpath's directory at the beginning",
	"    -a NAME modify path after given NAME",
	"    -b      put new arguments before existing path",
	"    -d      remove duplicates/non-directory items",
	"    -e      put new arguments after end of path",
	"    -f      allow filenames to match, as well as directories",
	"    -n NAME specify environment-variable to use (default: PATH)",
	"    -p      print in C-shell form",
	"    -r      remove arguments from path",
	"    -v      verbose",
	"",
	"Put a '-' before a command which is invoked with the environment variable",
	"updated, rather than echoing the result to standard output.",
	NULL
    };
    register int j;

    for (j = 0; tbl[j] != NULL; j++)
	fprintf(stderr, "%s\n", tbl[j]);
    exit(EXIT_FAILURE);
}

static char *
StrAlloc(const char *s)		/* patch: not everyone has 'strdup()' */
{
    char *d = malloc(strlen(s) + 1);

    if (d == NULL)
	failed("malloc");
    return strcpy(d, s);
}

static void
Remove(int offset, LIST * list, char *name)
{
    int n = offset;

    TRACE((stderr, "Remove %d:%s\n", offset, name));
    while (list[n].nn != NULL) {
	if (!Compare(list[n].nn, name)) {
	    do {
		list[n] = list[n + 1];
	    }
	    while (list[++n].nn != NULL);
	    n = offset - 1;
	}
	n++;
    }
}

static int
Append(int offset, LIST * list, char *name)
{
    int n = offset;
    LIST temp, save;

    TRACE((stderr, "Append %d:%s\n", offset, name));
    temp.nn = name;

    do {
	n++;
	save = list[n];
	list[n] = temp;
	temp = save;
    }
    while (list[n].nn != NULL);

    return offset + 1;
}

static int
exists(LIST * entry)
{
    TRACE((stderr, "exists(%s)\n", entry->nn));
    if (stat(entry->nn, &(entry->sb)) >= 0) {
	if ((entry->sb.st_mode & S_IFMT) == S_IFDIR) {
	    TRACE((stderr, "...directory\n"));
	    return TRUE;
	}
	if (allow_files
	    && (entry->sb.st_mode & S_IFMT) == S_IFREG) {
	    TRACE((stderr, "...file\n"));
	    return TRUE;
	}
    }
    TRACE((stderr, "...not found\n"));
    return FALSE;
}

#ifdef HAVE_REALPATH
static int
isFile(const char *path)
{
    struct stat sb;
    int result = FALSE;
    if (lstat(path, &sb) == 0
	&& (sb.st_mode & S_IFMT) == S_IFREG) {
	result = TRUE;
    }
    return result;
}

static int
isLink(const char *path)
{
    struct stat sb;
    int result = FALSE;
    if (lstat(path, &sb) == 0
	&& (sb.st_mode & S_IFMT) == S_IFLNK) {
	result = TRUE;
    }
    return result;
}
#endif

int
main(int argc, char *argv[])
{
    const char *name = "PATH";
    char out_delim = PATHDELIM;
    int remove_duplicates = FALSE;
    size_t length = (size_t) argc;
    int operation = 'a';
    int use_original = FALSE;
    const char *where = NULL;

    int c, point = 0;
    LIST *list;
    char *s;

    while ((c = getopt(argc, argv, "0a:bdefn:prv")) != EOF) {
	switch (c) {
	case '0':
	    use_original = TRUE;
	    break;
	case 'a':
	    where = optarg;
	    break;
	case 'b':
	    where = BLANK;
	    break;
	case 'd':
	    remove_duplicates = TRUE;
	    break;
	case 'e':
	    where = NULL;
	    break;
	case 'f':
	    allow_files = TRUE;
	    break;
	case 'n':
	    name = optarg;
	    break;
	case 'p':
	    out_delim = ' ';
	    break;
	case 'r':
	    if (where == NULL)
		where = BLANK;
	    operation = c;
	    break;
	case 'v':
	    out_delim = '\n';
	    break;
	default:
	    usage();
	}
    }

    /* Get the current path, make a corresponding list of strings.  Leave
     * enough room in the list to allow us to insert the arguments also.
     */
    if ((s = getenv(name)) == NULL)
	s = StrAlloc(BLANK);
    else
	s = StrAlloc(s);	/* ...just in case someone else uses it */

    for (c = 0; s[c] != EOS; c++)
	if (s[c] == PATHDELIM)
	    length++;
    length += (size_t) (argc + 3);
    list = (LIST *) calloc(length, sizeof(LIST));
    list[0].nn = StrAlloc(BLANK);	/* dummy entry, to simplify -b option */

    /* Split the environment variable into strings indexed in list[] */
    for (c = 1; *s != EOS; c++) {
	if (*s == PATHDELIM) {
	    list[c].nn = StrAlloc(".");
	} else {
	    list[c].nn = s;
	    while (*s != PATHDELIM && *s != EOS)
		s++;
	    if (*s == EOS)
		continue;
	    *s = EOS;
	}
	s++;
    }
    list[c].nn = NULL;
    TRACE((stderr, "%s has %d entries\n", name, c));

    /*
     * Find the directory where this program resides, accounting for symlinks.
     */
#ifdef HAVE_REALPATH
    if (use_original) {
	char *actual = malloc(PATH_MAX);
	if (actual != NULL) {
	    char *source = argv[0];
	    char *leaf;
	    int found = FALSE;
	    char resolved[PATH_MAX];
	    char which[PATH_MAX];

	    if (!isFile(source)) {
		for (c = 0; list[c].nn != NULL; ++c) {
		    size_t need = strlen(list[c].nn) + strlen(source);
		    if (need >= sizeof(which) - 3)
			continue;
		    sprintf(which, "%s/%s", list[c].nn, source);
		    if (isLink(which) || isFile(which)) {
			source = which;
			break;
		    }
		}
	    }

	    if (strchr(source, PATHSEP) != NULL
		&& realpath(source, actual) != NULL
		&& isFile(actual)
		&& (leaf = strrchr(actual, PATHSEP)) != NULL) {
		found = TRUE;
		*leaf = EOS;
	    } else if (strchr(source, PATHSEP) == NULL) {
		for (c = 0; list[c].nn != NULL; ++c) {
		    sprintf(actual, "%s%c%s", list[c].nn, PATHSEP, argv[0]);
		    while (isLink(actual)) {
			int have = (int) readlink(actual,
						  resolved,
						  sizeof(resolved));
			if (have <= 0)
			    break;
			resolved[have] = EOS;
			strcpy(actual, resolved);
		    }
		    if (isFile(actual)) {
			leaf = strrchr(actual, PATHSEP);
			if (leaf != NULL) {
			    found = TRUE;
			    *leaf = EOS;
			}
			break;
		    }
		}
	    }
	    if (found) {
		found = -1;
		for (c = 1; list[c].nn != NULL; ++c) {
		    if (!strcmp(list[c].nn, actual)) {
			found = c;
			break;
		    }
		}
#ifdef OPT_TRACE
		for (c = 1; list[c].nn != NULL; ++c) {
		    TRACE((stderr, " old %d %s\n", c, list[c].nn));
		}
#endif
		if (found > 1) {
		    /* rotate */
		    LIST inserts = list[found];
		    for (c = 1; c < found; ++c) {
			LIST pending = list[c];
			list[c] = inserts;
			TRACE((stderr, " new %d %s\n", c, list[c].nn));
			inserts = pending;
		    }
		    list[c] = inserts;
		    TRACE((stderr, " new %d %s\n", c, list[c].nn));
		} else if (found < 0) {
		    /* prepend */
		    LIST inserts;
		    LIST pending;
		    memset(&inserts, 0, sizeof(inserts));
		    inserts.nn = actual;
		    for (c = 1; list[c].nn != NULL; ++c) {
			pending = list[c];
			list[c] = inserts;
			TRACE((stderr, " new %d %s\n", c, list[c].nn));
			inserts = pending;
		    }
		    list[c] = inserts;
		    TRACE((stderr, " new %d %s\n", c, list[c].nn));
		}
	    }
	}
    }
#endif /* HAVE_REALPATH */

    /* Find the list-entry after which we insert/remove entries */
    for (c = 0; list[c].nn != NULL; c++) {
	if (where != NULL && !Compare(where, list[c].nn)) {
	    point = c;
	    break;
	}
    }
    if (list[c].nn == NULL)
	point = c - 1;
    TRACE((stderr, "argc=%d, point=%d (%c)\n", argc, point, operation));

    /* Perform the actual insertion/removal */
    while (optind < argc) {
	s = argv[optind++];
	if (!Compare(s, "-"))
	    break;
	if (operation == 'r') {
	    Remove(point, list, s);
	} else {
	    point = Append(point, list, s);
	}
    }

    if (remove_duplicates) {
	/* Check to see if the directory exists. If not, remove it */
	for (c = 1; list[c].nn != NULL; c++)
	    if (!exists(&list[c])) {
		Remove(c, list, list[c].nn);
		c--;
	    };
	/* Compare the inode & device numbers of the remaining items */
	for (c = 1; list[c].nn != NULL; c++) {
	    int d;

	    for (d = c + 1; list[d].nn != NULL; d++) {
		if (SameDir(&list[c], &list[d])) {
		    Remove(d, list, list[d].nn);
		    d--;
		}
	    }
	}
    }

    /* Finally, print the path */
    if (optind < argc) {
	size_t len = strlen(name) + 2;
	char *changed = NULL;

	for (c = 1; list[c].nn != NULL; c++)
	    len += 1 + strlen(list[c].nn);
	changed = malloc(len);
	strcpy(changed, name);
	for (c = 1; list[c].nn != NULL; c++) {
	    sprintf(changed + strlen(changed), "%c%s",
		    (c > 1) ? PATHDELIM : '=',
		    list[c].nn);
	}
	fflush(stderr);
	fflush(stdout);
	putenv(changed);
#ifdef WIN32
	spawnvp(_P_WAIT, argv[optind], &argv[optind]);
#else
	execvp(argv[optind], &argv[optind]);
#endif
	fflush(stderr);
	fflush(stdout);
    } else {
	fflush(stderr);
	for (c = 1; list[c].nn != NULL; c++) {
	    if (c > 1)
		(void) putchar(out_delim);
	    (void) fputs(list[c].nn, stdout);
	}
	(void) putchar('\n');
	fflush(stdout);
    }

#ifdef NO_LEAKS
    free(list);
#endif
    exit(EXIT_SUCCESS);
    /*NOTREACHED */
}
